testsuite/classlibrary/TestUGen_RTAlloc.sc

```diff
@@ -1,326 +1,325 @@
-			output = val[0];
-			cond.signalOne();
-		});
+	// Async helpers
 
-		cond.waitFor(1, { output != -1 });
-		this.assertFloatEquals(output, 0.0, "a failed RTAlloc should clear unit outputs")
-	}
-
-	test_allocFail_continueProcessing_sameNode {
-		var cond = CondVar(), output = -1;
-		{ DelayN.ar(Silent.ar, 100); DC.ar(1)}.loadToFloatArray(1 / server.sampleRate, server, {|val|
-			output = val[0];
+	awaitSynthOutput { |synthFn, runtimeSamples=1, timeout=1|
+		var cond = CondVar(), synthOutput = nil;
+		var runtimeSeconds = runtimeSamples / server.sampleRate;
+		// allow extra timeout for longer tests (e.g. testFFT with big memSize)
+		if (runtimeSeconds > timeout) { 
+			timeout = runtimeSeconds + timeout
+		};
+		synthFn.loadToFloatArray(runtimeSeconds, server) { |out|
+			synthOutput = out;
 			cond.signalOne();
-		});
-
-		cond.waitFor(1, { output != -1 });
-		this.assertFloatEquals(output, 1.0, "a failed RTAlloc should not block other units in the same node")
+		};
+		cond.waitFor(timeout) { synthOutput.notNil };
+		^synthOutput;
 	}
 
-	test_allocFail_continueProcessing_otherNodes {
-		var cond = CondVar(), output = -1;
-		server.makeBundle(nil) {
-			{ DelayN.ar(Silent.ar, 100) }.play(server);
-			{ DC.ar(1) }.loadToFloatArray(1 / server.sampleRate, server, {|val|
-				output = val[0];
-				cond.signalOne();
-			});
+	assertAllocPass { |name, func, runtimeSamples|
+		var out = this.awaitSynthOutput(func, runtimeSamples);
+		if (out.isNil) {
+			this.assert(false, "% allocPass test should complete".format(name))
+		} {
+			this.assert(out.sum != 0,
+				"% allocPass test should produce non-zero output".format(name)
+			);
 		};
-
-		cond.waitFor(1, { output != -1 });
-		this.assertFloatEquals(output, 1.0, "a failed RTAlloc should not block other nodes in the processing chain")
 	}
 
-	test_allocFail_setDoneFlag {
-		var cond = CondVar(), output = -1;
-		{
-			K2A.ar(Done.kr(DelayN.ar(Silent.ar, 100)))
-		}.loadToFloatArray(1 / server.sampleRate, server, {|val|
-			output = val[0];
-			cond.signalOne();
-		});
-
-		cond.waitFor(1, { output != -1 });
-		this.assertFloatEquals(output, 1.0, "a failed RTAlloc should set the Done flag")
+	assertAllocFail { |name, func, runtimeSamples|
+		var out = this.awaitSynthOutput(func, runtimeSamples);
+		if (out.isNil) {
+			this.assert(false, "% allocFail test should complete".format(name))
+		} {
+			this.assertEquals(out.sum, 0,
+				"% allocFail test should produce zero output".format(name)
+			);
+		};
 	}
 
-	// TEST ALL UGENS
-	// for each UGen in /server/plugins which calls RTAlloc
-	// test successful alloc: should output non-zero values
-	// test failed alloc: should only output zeros
+	// mem calc helpers
 
-	// FFT UGens needs separate testing (see below)
-	// Some UGens were left out (see end of file)
+	memSizeKb { ^server.options.memSize }
+	memSizeFloats { ^this.memSizeKb * (1024 / 4) }
+	memSizeSeconds { ^this.memSizeFloats / server.sampleRate }
```

Comment:
I wonder if these properties could be promoted to the `UnitTest` class for broad use across the test suite (and documented)?

Reply:
I think for now it makes sense to have them just in this test, since it's especially related to filling real-time memory. But if we have other cases where it could be useful we could think of moving these methods to UnitTest or even Server or ServerOptions.

Reply:
Sure, thanks for considering it!